File: packages/typescript/src/bootstrap/providers/knowledge.ts
import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "../../types/index.ts";
import { ModelType } from "../../types/index.ts";

/**
 * Knowledge Provider - Provides relevant knowledge from the agent's knowledge base.
 *
 * This provider retrieves and formats relevant knowledge entries
 * based on the current context and message using semantic similarity search.
 */
export const knowledgeProvider: Provider = {
  name: "KNOWLEDGE",
  position: 60,
  description:
    "Provides relevant knowledge from the agent's knowledge base based on semantic similarity",
  dynamic: true,

  get: async (runtime: IAgentRuntime, message: Memory, _state: State) => {
    const sections: string[] = [];
    const knowledgeEntries: Array<{
      id: string;
      text: string;
      source: string;
    }> = [];

    // Get message text for keyword search (if supported) or fallback
    const queryText = message.content?.text || "";

    if (!queryText) {
      return {
        text: "",
        values: {
          knowledgeCount: 0,
          hasKnowledge: false as boolean,
        },
        data: {
          entries: [],
          query: "",
        },
      } as ProviderResult;
    }

    // New Strategy: Use embeddings from recent messages to find relevant knowledge
    // This avoids blocking on generating an embedding for the CURRENT message

    // 1. Fetch recent messages
    const recentMessages = await runtime.getMemories({
      tableName: "messages",
      roomId: message.roomId,
      count: 5,
      unique: false,
    });

    // 2. Extract valid embeddings
    const embeddings = recentMessages
      .map((m) => m.embedding)
      .filter((e): e is number[] => !!e && e.length > 0);

    // If no recent embeddings (e.g. new conversation), we can't do vector search
    // without blocking. For now, we accept this trade-off for speed.
    // Ideally, we would fallback to keyword search if the adapter supports it.

    const relevantKnowledge = new Map<string, any>();

    if (embeddings.length > 0) {
      // 3. Search using recent embeddings
      // We'll use the most recent embedding primarily, or perhaps a few
      // For now, let's use the most recent one (last in the list likely, depending on sort)
      // recentMessages are usually returned sorted by time desc?
      // getMemories impl usually returns DESC.
      // So the first one is the most recent.

      const primaryEmbedding = embeddings[0];

      const results = await runtime.searchMemories({
        tableName: "knowledge",
        embedding: primaryEmbedding,
        query: queryText, // Pass query text for hybrid search if supported
        count: 5,
        match_threshold: 0.75, // Reasonable threshold
      });

      for (const entry of results) {
        if (entry.id) relevantKnowledge.set(entry.id.toString(), entry);
      }
    }

    // 4. Format results
    const uniqueEntries = Array.from(relevantKnowledge.values());

    for (const entry of uniqueEntries) {
      if (entry.content?.text) {
        let knowledgeText = entry.content.text;
        // Truncate if too long
        if (knowledgeText.length > 500) {
          knowledgeText = `${knowledgeText.substring(0, 500)}...`;
        }

        const entryData = {
          id: entry.id?.toString() || "",
          text: knowledgeText,
          source: (entry.metadata?.source as string | undefined) || "unknown",
        };
        knowledgeEntries.push(entryData);
        sections.push(`- ${knowledgeText}`);
      }
    }

    const contextText =
      sections.length > 0 ? `# Relevant Knowledge\n${sections.join("\n")}` : "";

    return {
      text: contextText,
      values: {
        knowledgeCount: knowledgeEntries.length,
        hasKnowledge: knowledgeEntries.length > 0,
      },
      data: {
        entries: knowledgeEntries,
        query: queryText,
      },
    } as ProviderResult;
  },
};

File: packages/rust/src/bootstrap/providers/knowledge.rs
//! KNOWLEDGE provider implementation.

use async_trait::async_trait;

use crate::error::PluginResult;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

/// Provider for knowledge information.
pub struct KnowledgeProvider;

#[async_trait]
impl Provider for KnowledgeProvider {
    fn name(&self) -> &'static str {
        "KNOWLEDGE"
    }

    fn description(&self) -> &'static str {
        "Provides relevant knowledge from the agent's knowledge base based on semantic similarity"
    }

    fn is_dynamic(&self) -> bool {
        true
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let query_text = &message.content.text;

        if query_text.is_empty() {
            return Ok(ProviderResult::new("")
                .with_value("knowledgeCount", 0)
                .with_value("hasKnowledge", false)
                .with_data("entries", Vec::<serde_json::Value>::new()));
        }

        let mut sections = Vec::new();
        let mut knowledge_entries: Vec<serde_json::Value> = Vec::new();

        // Search for relevant knowledge
        match runtime.search_knowledge(query_text, 5).await {
            Ok(entries) => {
                for entry in entries {
                    if !entry.content.text.is_empty() {
                        let mut knowledge_text = entry.content.text.clone();
                        if knowledge_text.len() > 500 {
                            knowledge_text.truncate(500);
                            knowledge_text.push_str("...");
                        }

                        let source = entry
                            .metadata
                            .get("source")
                            .and_then(|v| v.as_str())
                            .unwrap_or("unknown")
                            .to_string();

                        knowledge_entries.push(serde_json::json!({
                            "id": entry.id.to_string(),
                            "text": knowledge_text,
                            "source": source
                        }));

                        sections.push(format!("- {}", knowledge_text));
                    }
                }
            }
            Err(e) => {
                runtime.log_warning(
                    "provider:knowledge",
                    &format!("Error searching knowledge: {}", e),
                );
            }
        }

        let context_text = if sections.is_empty() {
            String::new()
        } else {
            format!("# Relevant Knowledge\n{}", sections.join("\n"))
        };

        Ok(ProviderResult::new(context_text)
            .with_value("knowledgeCount", knowledge_entries.len() as i64)
            .with_value("hasKnowledge", !knowledge_entries.is_empty())
            .with_data("entries", knowledge_entries)
            .with_data("query", query_text.clone()))
    }
}
